rust/hyperlane-base/src/types/gcs_storage.rs

```diff
@@ -0,0 +1,245 @@
+use crate::CheckpointSyncer;
+use async_trait::async_trait;
+use derive_new::new;
+use eyre::{bail, Result};
+use hyperlane_core::{SignedAnnouncement, SignedCheckpointWithMessageId};
+use std::fmt;
+use ya_gcp::{
+    storage::StorageClient, AuthFlow, ClientBuilder, ClientBuilderConfig, ServiceAccountAuth,
+};
+
+const LATEST_INDEX_KEY: &str = "gcsLatestIndexKey";
+const ANNOUNCEMENT_KEY: &str = "gcsAnnouncementKey";
+/// Path to GCS users_secret file
+pub const GCS_USER_SECRET: &str = "GCS_USER_SECRET";
+/// Path to GCS Service account key
+pub const GCS_SERVICE_ACCOUNT_KEY: &str = "GCS_SERVICE_ACCOUNT_KEY";
+
+/// Google Cloud Storage client builder
+/// Provide `(None, None)` for no-auth access to public bucket[s]
+/// # Example 1 - anonymous client with access to public bucket[s]
+/// ```
+///    use hyperlane_base::GcsStorageClientBuilder;
+/// #  #[tokio::main]
+/// #  async fn main() {
+///    let client = GcsStorageClientBuilder::new(None, None)
+///        .build("HyperlaneBucket", None)
+///        .await.expect("failed to instantiate anonymous client");
+/// #  }
+///```
+///
+/// For authenticated write access to bucket[s] proper file path must be provided.
+/// # WARN: panic-s if file path is incorrect or data in it as faulty
+///
+/// # Example 2 - service account key
+/// ```should_panic
+///    use hyperlane_base::GcsStorageClientBuilder;
+/// #  #[tokio::main]
+/// #  async fn main() {
+///    let client = GcsStorageClientBuilder::new(Some("path/to/sac.json".into()), None)
+///        .build("HyperlaneBucket", None)
+///        .await.expect("failed to instantiate anonymous client");
+/// #  }
+///```
+/// # Example 3 - user secret access
+/// ```should_panic
+///    use hyperlane_base::GcsStorageClientBuilder;
+/// #  #[tokio::main]
+/// #  async fn main() {
+///    let client = GcsStorageClientBuilder::new(None, Some("path/to/user_secret.json".into()))
+///        .build("HyperlaneBucket", None)
+///        .await.expect("failed to instantiate anonymous client");
+/// #  }
+///```
+#[derive(Debug, new)]
+pub struct GcsStorageClientBuilder {
+    /// A path to the oauth service account key json file.
+    ///
+    /// If this is not provided, `user_secrets` must be.
+    service_account_key: Option<String>,
+
+    /// Path to oauth user secrets, like those created by
+    /// `gcloud auth application-default login`
+    ///
+    /// If this is not provided, `service_account_key` must be.
+    user_secrets: Option<String>,
+}
+
+/// Google Cloud Storage client
+/// Enables use of any of service account key OR user secrets to authenticate
+/// For anonymous access to public data provide `(None, None)` to Builder
+pub struct GcsStorageClient {
+    // GCS storage client
+    // # Read methods:
+    //
+    // * `get_object(&self,
```

Comment:
Since this comment is just a copy of the ya-gcp docs entry, I don't think it's required here. Additionally, these methods may change in a future version and the compiler will complain if that's the case, whereas this comment will become stale

Reply:
removed. only ref url left for quick access to the documentation.